src/main/java/io/jenkins/plugins/adoptopenjdk/AdoptOpenJDKInstaller.java

```diff
@@ -105,26 +115,52 @@ public FilePath performInstallation(ToolInstallation tool, Node node, TaskListen
-            FilePath installation = zipExtractionInstaller.performInstallation(tool, node, log);
-
-            installation.child(".timestamp").delete(); // we don't use the timestamp
-            FilePath base = findPullUpDirectory(installation, p);
-            if (base != null && base != expected) {
-                base.moveAllChildrenTo(expected);
+            File cache = getLocalCacheFile(p, c);
+            if (!DISABLE_CACHE && cache.exists() && cache.length()>1*1024*1024) { // if the file is too small, don't trust it
+                // the zip contains already the directory so we unzip to parent directory
+                expected.getParent().installIfNecessaryFrom(cache.toURI().toURL(), log,
+                                                            Messages.AdoptOpenJDKInstaller_performInstallation_path(expected));
+                expected.getParent().child( ".timestamp" ).delete();
+            } else {
+                String url = binary.binary_link;
+                ZipExtractionInstaller zipExtractionInstaller = new ZipExtractionInstaller(null, url, null);
+                FilePath installation = zipExtractionInstaller.performInstallation(tool, node, log);
+                installation.child(".timestamp").delete(); // we don't use the timestamp
+                FilePath base = findPullUpDirectory(installation, p);
+                if (base != null && base != expected) {
+                    base.moveAllChildrenTo(expected);
+                }
+                marker.write(id, null);
+                if(!DISABLE_CACHE) {
+                    // update the local cache on master
+                    // download to a temporary file and rename it in to handle concurrency and failure correctly,
+                    Path tmp = new File( cache.getPath()+".tmp").toPath();
+                    try {
+                        Path tmpParent = tmp.getParent();
+                        if (tmpParent != null) {
+                            Files.createDirectories(tmpParent);
+                        }
+                        try (OutputStream out = Files.newOutputStream(tmp)) {
+                            expected.zip(out);
+                        }
+                        Files.move(tmp, cache.toPath(), StandardCopyOption.REPLACE_EXISTING);
+                    } finally {
+                        Files.deleteIfExists(tmp);
+                    }
+                }
             }
-            marker.write(id, null);
-
-            return installation;
         } catch (DetectionFailedException e) {
             log.getLogger().println(Messages.AdoptOpenJDKInstaller_performInstallation_JdkSkipped(e.getMessage()));
         }
 
         return expected;
     }
 
+    private File getLocalCacheFile( Platform platform, CPU cpu) {
+        // we force .zip file
+        return new File(Jenkins.get().getRootDir(), "caches/adoptopenjdk/"+platform+"/"+cpu+"/"+id+".zip");
```

Comment:
There is no standard for a standard caches location inside Jenkins, but many tools already use this location